content/cpi.md

```diff
@@ -0,0 +1,339 @@
+
+# TL;DR
+
+- A **Cross-Program Invocation (CPI)** is a call from one program to another, targeting a specific instruction on the called program.
+- CPIs are made using the commands **`invoke`** or **`invoke_signed`**, which is also how programs provide signatures for PDAs that they own.
+- CPIs make all programs in the Solana ecosystem completely interoperable because all public instructions of a program can be invoked by another program via a CPI.
+- Because we have no control over the accounts and data submitted to a program, it's important to verify all of the parameters passed into a CPI to ensure program security.
+
+# Overview
+
+## What is a CPI?
+
+A Cross-Program Invocation is when one program calls an instruction on another program. Cross-program invocations allow programs to invoke other programs directly, but the depth is constrained currently to 4. Remember very early in the course when we talked about how all programs are callable from any client? Well, the same is true for programs on chain! Any on-chain program can make calls to any other program’s public instructions, they just need to construct the instruction correctly. You can make CPIs to native programs, other programs you’ve created, as well as third party programs. CPIs essentially turn the entire Solana ecosystem into one giant API that is at your disposal as a developer.
+
+CPIs have a similar make up to instructions that you are used to creating client side. There are some intricacies and differences depending on if you are using `invoke` or `invoke_signed`, both of which we will cover later in the lesson.
+
+## How to make a CPI
+
+CPIs are made using the [`invoke`](https://docs.rs/solana-program/1.10.19/solana_program/program/fn.invoke.html) or [`invoke_signed`](https://docs.rs/solana-program/1.10.19/solana_program/program/fn.invoke_signed.html) function from the `solana_program` crate. The former is what’s used when the signatures needed for the transaction are *not* for PDAs. The latter is what’s used when a program needs to provide a signature for a PDA.
+
+```rust
+// used when there are not signatures for PDAs needed
+pub fn invoke(
+    instruction: &Instruction,
+    account_infos: &[AccountInfo<'_>]
+) -> ProgramResult
+
+// used when a program must provide a 'signature' for a PDA, hence the signer_seeds parameter
+pub fn invoke_signed(
+    instruction: &Instruction,
+    account_infos: &[AccountInfo<'_>],
+    signers_seeds: &[&[&[u8]]]
+) -> ProgramResult
+```
+
+CPIs extend the privileges of the caller to the callee. Meaning that any account that is passed in to your program as a signer or writable account will also be considered a signer or writable account inside the program being called. For example, if the instruction the caller is processing contains a signer or writable account, then the caller can invoke an instruction that also contains that signer and/or writable account. Actually using `invoke` to make a CPI may look something like this:
```

Comment:
Yes, you basically have to build the instruction all over again - so not every account that is passed into the original program will be passed along to the program invoked by the cpi. You have to explicitly include which accounts you want to pass into the cpi.

Reply:
@ixmorrow cool that's what I thought. Can we make that clearer here? It reads a little bit like all accounts passed to the first program will be passed to the second program.

Reply:
I re-worded this, kinda sounds like a mouthful but does it make more sense to you?

"CPIs extend the privileges of the caller to the callee. Meaning that any account that is passed into your program as a signer or writable account, if it is then passed into a CPI, will also be considered a signer or writable account inside the program being invoked. For example, if the instruction the callee program is processing contains an account that was marked as a signer or writable when originally passed into the caller program, then it will be considered a signer or writable account in the invoked program as well.

It's important to note that you as the developer decide which accounts to pass into the CPI. You can think of a CPI as building another instruction from scratch with only information that was passed into your program."